From 0442f2b851b4280819d165cd94bcec2fc80f1c52 Mon Sep 17 00:00:00 2001
From: Bradley Lowekamp <blowekamp@mail.nih.gov>
Date: Wed, 29 Dec 2021 18:10:12 +0000
Subject: [PATCH 2/2] ENH: remove rendering datetime with Doxygen into HTML
 footer

The date and time are not set by javascript when the document is
loaded. This removes nightly modification of all the HTML files, and
enable traditional static HTML cache to work, and enable the generated
HTML to be committed to git regularly without nightly modification.
---
 Documentation/Doxygen/DoxygenFooter.html |  3 ++-
 Documentation/Doxygen/DoxygenHeader.html |  1 +
 Utilities/Doxygen/CMakeLists.txt         | 10 ++++++++++
 Utilities/Doxygen/build_text.js.in       |  3 +++
 Utilities/Doxygen/datetime.py            | 21 +++++++++++++++++++++
 5 files changed, 37 insertions(+), 1 deletion(-)
 create mode 100644 Utilities/Doxygen/build_text.js.in
 create mode 100755 Utilities/Doxygen/datetime.py

diff --git a/Documentation/Doxygen/DoxygenFooter.html b/Documentation/Doxygen/DoxygenFooter.html
index 7b8abb255e..b4fd416c04 100644
--- a/Documentation/Doxygen/DoxygenFooter.html
+++ b/Documentation/Doxygen/DoxygenFooter.html
@@ -21,7 +21,8 @@
   </small>
 </div>
 <address class="footer"><small>
-$generatedby &#160;<a href="http://www.doxygen.org/index.html">
+Generated on <span id="datetime">unknown</span> for $projectname by &#160;
+<a href="http://www.doxygen.org/index.html">
 <img class="footer" src="$relpath^doxygen.png" alt="doxygen"/>
 </a> $doxygenversion
 </small></address>
diff --git a/Documentation/Doxygen/DoxygenHeader.html b/Documentation/Doxygen/DoxygenHeader.html
index b7bb84d8ca..ccd5f16303 100644
--- a/Documentation/Doxygen/DoxygenHeader.html
+++ b/Documentation/Doxygen/DoxygenHeader.html
@@ -14,6 +14,7 @@
 $treeview
 $search
 $mathjax
+<script type="text/javascript" src="$relpath^build_text.js"></script>
 <link href="$relpath^$stylesheet" rel="stylesheet" type="text/css" />
 $extrastylesheet
 </head>
diff --git a/Utilities/Doxygen/CMakeLists.txt b/Utilities/Doxygen/CMakeLists.txt
index 61076312b0..e7f49e358e 100644
--- a/Utilities/Doxygen/CMakeLists.txt
+++ b/Utilities/Doxygen/CMakeLists.txt
@@ -140,6 +140,16 @@ if (ITK_BUILD_DOCUMENTATION)
   file(COPY workbox-cli-config.js serviceWorker.js.in
     DESTINATION ${ITK_BINARY_DIR}/Utilities/Doxygen/)
 
+  execute_process(COMMAND ${PYTHON_EXECUTABLE} "${CMAKE_CURRENT_LIST_DIR}/datetime.py"
+     RESULT_VARIABLE CMD_RESULT
+     OUTPUT_VARIABLE _DATETIME)
+
+   if (CMD_RESULT)
+     message(FATAL_ERROR "Datetime failed!")
+   endif()
+
+   configure_file(${CMAKE_CURRENT_LIST_DIR}/build_text.js.in
+     "${ITK_DOXYGEN_OUTPUT_DIR}/html/build_text.js")
 
   find_package( LATEX )
 
diff --git a/Utilities/Doxygen/build_text.js.in b/Utilities/Doxygen/build_text.js.in
new file mode 100644
index 0000000000..bd0bef0a5e
--- /dev/null
+++ b/Utilities/Doxygen/build_text.js.in
@@ -0,0 +1,3 @@
+$(function(){
+    document.getElementById("datetime").textContent = "@_DATETIME@"
+});
diff --git a/Utilities/Doxygen/datetime.py b/Utilities/Doxygen/datetime.py
new file mode 100755
index 0000000000..a6e7953733
--- /dev/null
+++ b/Utilities/Doxygen/datetime.py
@@ -0,0 +1,21 @@
+#!/usr/bin/env python
+#
+#  Copyright NumFOCUS
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0.txt
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+# Writes the date and time in RFC 2822 format to stdout
+
+import  time
+import sys
+sys.stdout.write(time.strftime('%a, %d %b %Y %H:%M:%S +0000', time.gmtime()))
-- 
2.25.1

